lessons/bash-intro/lesson.md

```diff
@@ -0,0 +1,330 @@
+
+    [directory-name/]filename
+
+Except for a few cases, the pathname is relative to the *current working directory*. So if you are in your home directory, `Desktop/my_script.py` refers to the `my_script.py` file in your own Desktop folder, not anyone else's. If you're in your `Desktop` folder, then `Desktop/my_script.py` refers to a file inside the Desktop folder inside your Desktop folder (which probably isn't there), but `my_script.py` refers to the previous file.
+
+Try it out: from your home directory, run the following commands (`touch` creates a file):
+
+```
+mkdir tmp
+touch tmp/file_a
+ls
+```
+
+In the list of files, you should see the `tmp` directory, but no `file_a`. Now execute
+
+```
+ls tmp
+```
+
+you should see the files in the `tmp` directory (`file_a`). Now try
+
+```
+cat tmp/file_a
+cd tmp
+cat tmp/file_a
+```
+
+What went wrong?
+
+
+## Absolute paths
+
+Sometimes, you'll need to access files from another location in the file hierarchy. With relative paths, you can use `..` to access the parent directory, so `../../src/include/` goes up two directories, then down through `src` and `include`. This can get cumbersome if you're not sure exactly how many times you need to go up. 
+
+Instead, there are absolute paths. On Unix systems, there is one root node for the file-system: `/` (compare this with `C:\` in Windows). All files are in subdirectories of root. Any pathname starting with `/` is an absolute path starting from the root.
+
+Try out `ls /`
+
+Another special location is your home directory. The short-form is `~/`. On a Mac, this would be `/Users/username/`. You can also short-form any user's home-directory with `~username/`.
+
+Try `cat ~/.bashrc`.
+
+Most of the time working with scripts, you should use relative paths, because one cannot assume the state of another's file system.
+
+
+# Globbing
+
+If you want to pass more than one filename as an argument, you can use *globbing*. The special patterns for globbing are:
+
+- `*` Match any string of arbitrary length
+- `?` Match any single character
+- `[]` Match any of the characters within square brackets.
+
+Try moving all python files into the `src` directory. How about copying every file with `a`?
+
+# Getting more powerful
+
+Some true strengths of the command line come from combining commands. This is the "Unix philosophy" of having many small programs that perform their jobs well. These units can be combined to perform more complicated tasks.
+
+# Pipes (stdin/out)
+
+Commands can be strung together with pipes `|`. Those of you using R may have seen the concept in the `dplyr` package. Commands typically have an input (stdin) and an output (stdout). The pipe attaches the output of one command into the input of another.
+
+This is easier to see with commands like `cat` and `head`. Try these out using our demo directory:
+
+```
+head animals
+cat animals
+cat animals | head
+ls | head -n2
+head
```

Comment:
A tricky one, because it's just going to hang waiting for input (and you haven't mentioned Ctrl+C or Ctrl+D.)